prql-java/prql-api/src/main/java/org/prqllang/internal/Utils.java

```diff
@@ -0,0 +1,247 @@
+            tryLoadLibrary(userLibName, true);
+            return; // NOSONAR
+        } catch (Throwable t) { // NOSONAR
+            if (debug) {
+                throw t;
+            }
+        }
+
+        final String packedLibName = new StringBuilder(NATIVE_RESOURCE_HOME).append(path).append('/').append(fileName)
+                .toString();
+        URL url = Utils.class.getResource(packedLibName);
+        final String tmpLibName;
+        if (url == null) {
+            throw new UnsatisfiedLinkError(String.format(Locale.ROOT, "Failed to load native libraries: %s and %s",
+                    userLibName, packedLibName));
+        } else {
+            File file = new File(userLibName);
+            if (file.isDirectory() || (file.exists() && !file.canWrite())) {
+                file = Paths.get(System.getProperty("tmp.dir"), "prql", fileName).toFile();
+                if (!file.exists()) {
+                    file.mkdirs();
+                }
+                tmpLibName = file.getAbsolutePath();
+            } else {
+                tmpLibName = userLibName;
+            }
+        }
+
+        byte[] bytes = new byte[4096];
+        try (InputStream in = url.openStream(); OutputStream out = new FileOutputStream(tmpLibName, false)) {
+            int read = 0;
+            while ((read = in.read(bytes)) != -1) {
+                out.write(bytes, 0, read);
+            }
+            out.flush();
+        } catch (IOException e) {
+            if (debug) {
+                throw new IllegalStateException(e);
+            }
+
+            String errorMsg = e.getMessage();
+            try {
+                errorMsg = String.format(Locale.ROOT, "Failed to export library %s from %s due to: %s", packedLibName,
+                        Paths.get(Utils.class.getProtectionDomain().getCodeSource().getLocation().toURI())
+                                .toFile().getName(),
+                        errorMsg);
+            } catch (URISyntaxException exp) {
+                if (debug) {
+                    throw new IllegalStateException(exp);
+                }
+
+                errorMsg = String.format(Locale.ROOT, "Failed to export library %s due to: %s", packedLibName,
+                        errorMsg);
+            }
+            throw new UnsatisfiedLinkError(errorMsg);
+        }
+
+        tryLoadLibrary(tmpLibName, true);
+    }
+
+    /**
+     * Loads native library.
+     *
+     * @param lib      non-null name of the native library, or an absolute path to
+     *                 the library
+     * @param absolute whether load the library from absolute path
+     */
+    static void tryLoadLibrary(String lib, boolean absolute) {
+        if (absolute) {
+            System.load(lib);
+        } else {
+            System.loadLibrary(lib);
+        }
+    }
+
+    // https://github.com/netty/netty/blob/5ce448951c46972d58765b3be8684f589133b0a3/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L138
```

Comment:
This was copied from Netty. I only put reference link at the top. Is there any specific format that I need to follow? Am I supposed to update files like NOTICE or so?

Reply:
As long as there is a link and the license is compatible (Netty's is) then I _think_ it's fine. 

Someone should correct me if we need to do more; e.g. copy the license.